flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java

```diff
@@ -260,17 +258,28 @@ protected StreamTask(
 			@Nullable TimerService timerService,
 			Thread.UncaughtExceptionHandler uncaughtExceptionHandler,
 			StreamTaskActionExecutor actionExecutor,
-			TaskMailbox mailbox) {
+			TaskMailbox mailbox) throws Exception {
 
```

Comment:
If I'm not mistaken, these exceptions were previously thrown from `StreamTask.invoke()`; and then they weren't wrapped into anything neither by `StreamTask` nor by `Task`.

Now they are thrown from `Task.loadAndInstantiateInvokable` which can wrap them into `FlinkException`. 
This wrapping can break exception-matching logic in `Task` (e.g. `if (t instanceof CancelTaskException) ...`).

Reply:
Good catch. I double checked the codes and it seems not break any existing behaviors, because we would not explicitly match the `FlinkException` in `Task` or `JobMaster` ATM.

This wrapped exception for `statelessCtor.newInstance(environment)` is redundant before this PR, because any instance of `AbstractInvokable` would not throw exceptions during construction.

I am a bit torn whether it should be wrapped into `FlinkException` now because it would hide the essential exception to confuse users sometimes. But this is a common issue also existing in other wrapped exceptions.  E.g. some users were always confused when seeing the message "Could not forward element to next operator" from wrapped `ExceptionInChainedOperatorException` which causes task failure, but the real exception might come from specific UDF. And I also plan to improve this issue a bit by appending the real exception message with the custom message, but it should be a separate ticket. 

Do you have some further concerns for this case?

Reply:
Hmm..I meant that we could match something other than `FlinkException`, but it is wrapped into `FlinkException`.

I checked it again and found only one such case: `c` on line 776 (`Task`).
So if it is thrown somewhere in constructor of `StreamTask` or its descendants then `Task` will be failed, not cancelled. Currently, it is not thrown in any of constructors though.

I think we can ignore this issue for now. And if `CancelTaskException` will be added to constructor, IMO we should rather revisit this cancelation mechanism.